### ISSUE TITLE
Blurry fonts in high DPI

### ISSUE THREAD

**Issue by acterhd** [ocornut/imgui#1065]
Hello. I tried use scaling in IMGui but seen that font too blurry.

**Comment by meshula** [comment]
You have to ask for double-sized fonts. Instead of asking for a size of 10, ask for 20, etc.

**Comment by drewish** [comment]
Yeah I'd hoped that `IO.FontGlobalScale` would be applied when the font's texture is built, but it's applied afterwards and scales the texture. You can't rely on the default font since you'll need to pass in a size.

**Comment by Matheus28** [comment]
A workaround for this is to add a different font other than the default and setting `OversampleH` and `OversampleH` to something high, I use 3 for both.

**Comment by slajerek** [comment]
This is still not resolved. I literally tried everything, default stb backend gives blurry fonts. FreeType with oversampling - does not work, as FreeType does not support oversampling. How to make fonts look good on macOS with High DPI?

**Comment by thedmd** [comment]
You don't need to change your font size, which will screw your layout. What you need is to increase resolution of generated font atlas. And that is what `ImFontConfig::Density` is for.

You can cherry pick patch to ImGui & FreeType atlas builder that add `ImFontConfig::Density` support from my
[feature/hidpi-support](https://github.com/thedmd/imgui/tree/feature/hidpi-support) branch.

Note: Oversampling afaik is here to fight artifacts on small fonts rendered with stb_truetype. It it not suited to change atlas density.

| MacBook | iPad |
|-|-|
|<img width="1497" alt="MacBook@2" src="https://user-images.githubusercontent.com/1197433/147016339-e55b574c-909c-4fdd-9235-2d959f719f8f.png">|<img width="834" alt="iPad@2" src="https://user-images.githubusercontent.com/1197433/147015597-c886f04c-cbe3-4213-b726-1ba2b8cd910d.png">|

**Comment by whitequark** [comment]
@thedmd Do you have any plans to upstream the font density patch? (If not I might pick it up myself if it works well for me.)

**Comment by thedmd** [comment]
@whitequark Ability to control density was merged into ImGui about 8 months ago. I forgot about this thread, so I will take the opportunity now to mention it. :)

You can control density with `RasterizationDensity`.

https://github.com/ocornut/imgui/blob/1ec1f7a3de2ce96bee58b2825917707cc1b7c17c/imgui.h#L3005

**Comment by whitequark** [comment]
Thanks! This has sent me to a 7-hour-long rabbit hole of [updating the Rust bindings...](https://github.com/imgui-rs/imgui-rs/pull/787).

**Comment by jdchathuranga** [comment]
Seems the apple metal demo project text not supporting HDPI font scaling. The problem is the library is using the default font, for some reason DPI scaling is not working with the default font.
If you look at the end of ```-(instancetype)initWithNibName:(nullable NSString *)nibNameOrNil bundle:(nullable NSBundle *)nibBundleOrNil``` function in the `main.mm` file, you can see commented code for loading custom fonts.
uncomment one of these lines and give the correct path for the font. ```io.Fonts->AddFontFromFileTTF("/Users/abc/projects/imgui-docking/misc/fonts/DroidSans.ttf");``` this should fix the font problem and would look gorgeous.  (I couldn't get the relative path working, but the absolute path for the font worked, I think that is a my issue)

<img width="900" alt="Image" src="https://github.com/user-attachments/assets/315bc2f9-15d8-4ccc-8736-fe2e53e9c81a" />
